### DISCUSSION TITLE
What if we looked at the dependency tree as a more broader graph instead of a tree?

### DISCUSSION THREAD

**Discussion by shubhbapna** [kiegroup/build-chain-configuration-reader#87] [Ideas]
Currently, given a dependency file we produce a project tree to map relations between projects.
We use this mapping to:
- Generate the list of upstream projects for a given project
- Generate the list of full downstream projects for a given project

Now consider the following dependency file: [droolsjbmp-build-bootstrap](https://github.com/kiegroup/droolsjbpm-build-bootstrap/blob/main/.ci/project-dependencies.yaml)

Notice some issues that make the tree implementation difficult:
- duplicate edges:
   - appformer depends on lienzo-core, droolsjbpm-build-bootstrap, lienzo-tests, kiegroup/kie-soup even though to build lienzo-tests and kie-soup we would have had to build lienzo-core and droolsjbpm-build-bootstrap anyways. So we can simplify appformer dependencies to lienzo-tests and kie-soup.
   - Since we maintain a list of parents and children node in our tree implementation these duplicates increase the number of parents for appformer and number of children for lienzo-core, droolsjbpm-build-bootstrap
- Cycle: it has a cycle (appformer -> lienzo-tests -> lienzo-core -> appformer). But trees shouldn't have cycles
- Our current tree structure needs to store parent and children. This produces a circular json which comes with it's own set of problems
- To produce the list, we traverse the tree, find the leaves first then again traverse the tree etc etc

Proposed new graph structure:
- For each project we will create a node where we store the project name, array of outgoing edges, array of incoming edges
  - Array of outgoing edges:
    -  It is just an array of project names that are required to build the given node
  - Array of incoming edges:
    -  It is just an array of project names that depend on the given node

How can graph help?
- No circular json! Just need to store the project names.
- Duplicate edges won't matter
- Cycles are okay in graphs
- We just have to run DFS :)
- Maintenance becomes easier

**Comment by Ginxo** [comment]
I agree, I'm understanding project name will be treated as the node identifier, right?
Thanks for the proposal, I think this would be a great achievement

**Reply by shubhbapna** [reply]
Yep!

**Comment by Ginxo** [comment]
solved by https://github.com/kiegroup/build-chain-configuration-reader/issues/88
